### PR TITLE
DDNS Forward/Reverse updates

### PR DESCRIPTION
**Pull request by leshow**
#10 

involves RFC 4701/4702/4703.

I had thought about using the `kea-dhcp-ddns` binary to do the actual update, so there is some test code there for calling that. However, after actually starting to implement, most of the work was in dora itself. You need to still create the DHCID, & handle fqdn flags, for instance. The binary just creates and sends DNS updates, which we can use trust-dns-client for potentially.

I've been referencing kea's implementation.

### CODE REVIEW

**Comment by leshow** [comment]
The goal of this PR is to implement DDNS updates as specified in 4701/4702/4703

[4701](https://www.rfc-editor.org/rfc/rfc4701) <- DHCID rdata
[4702](https://www.rfc-editor.org/rfc/rfc4702) <- Client FQDN option section 4 specifies "server behavior" based on fqdn flags
[4703](https://www.rfc-editor.org/rfc/rfc4703) <- specifies the DHCID RR record that must be included in the DNS update

It's worth looking at Kea's DDNS docs also [here](https://kea.readthedocs.io/en/kea-2.0.0/arm/ddns.html#overview)

I propose adding the following section to the dora `config.yaml` file. It should be able to be specified per-range/static reservation and globally. When a lease is handed out, dora will look for the ddns config with precedence given to the range or reservation first.

```
ddns:
    # send updates. If the ddns section header is defined, enable_updates defaults to true
    enable_updates: true
    # default false. whether to override the client update FQDN flags
    override_client_updates: false
    # default false. whether to override the no update FQDN flags
    override_no_updates: false
    # optional, appends this suffix to the hostname option (opt 12)
    hostname_suffix: "text.suffix."
    # list of forward DNS servers
    forward: 
       - key: "key_foo" # optional, must match key name in tsig_keys
         ip: 192.168.3.111
    reverse:
       - key: "key_foo" # optional
         ip: 192.168.3.111
    # map of tsig keys. DNS servers reference these by name
    tsig_keys:
        key_foo:
          algorithm: "hmac-sha1"
          data: "<keydata>"
```

There are a few options that can change the behavior of the update in this PR. These options are similar to what is available in kea.

`enable_updates`: should we process the client FQDN option? true/false
`override_client_updates`: the client FQDN flag can have a flag telling the server that it wants to do the DNS update, setting this to true will _override_ that behavior and send back the relevant 'o' flag set to true. (see here: https://www.rfc-editor.org/rfc/rfc4702.html#section-4)
`override_no_updates`: client FQDN flags can have a 'no update' flag set, if `override_no_updates` is true, then we will do the update anyway and set the override flag on response.

The logic for flag handling is largely in the `handle_flags` function, and was mostly translated from Kea. [Keas flag handling](https://github.com/isc-projects/kea/blob/9c76b9a9e55b49ea407531b64783f6ec12546f42/src/lib/dhcpsrv/d2_client_mgr.cc#L115)

Left for a future task, is further interaction with option 12, the hostname option. The current PR will attempt a DNS update if the client FQDN option is not present but hostname option is. It seems as though different implementations handle that option in different ways, what is implemented so far in dora is the ability to append a suffix to the client provided hostname that will be used for the DNS update (`hostname_suffix`).

Kea's ddns config provides also a `name` field for both forward & reverse servers that it searches through, doing an update for whichever server `name` matches the client FQDN. In this proposal there is no such matching logic. If we allow the `ddns` config to be nested in the range & reservation blocks, I don't think we need to implement any matching logic, at least for now. It will work simply as: if a `ddns` block is present for a range, the flags and config determine that a DNS update should be made, then dora will send an update to every forward & reverse server in the list. Alternatively, one could just put a global ddns block and all updates would be directed to the servers specified.

As for the content of the DNS updates themselves, here is an example of a forward update

fwd_update

The DHCID RR is created in accordance with [4701](https://www.rfc-editor.org/rfc/rfc4701#section-3.5).


There is some integration work left to do on this PR still, and of course testing.

**Comment by astro** [comment]
What is missing for this feature to get merged?

**Comment by leshow** [comment]
Sorry @astro I didn't see your comment. I sort of hacked this together over xmas, it has some rough edges that need to be removed and code conflicts resolved. I should also probably run it through some testing in GNS3 to see that things are working properly.

I was basically waiting for someone to comment and say whether the config and behavior was acceptable and/or was missing anything they needed.

Do you have any comments on the configuration format or about the pcaps I posted?

File: example.yaml [hunk 1]
```diff
@@ -180,4 +178,30 @@ v6:
...
+    enable_updates: true
+    # default false. whether to override the client update FQDN flags
+    override_client_updates: false
+    # default false. whether to override the no update FQDN flags
+    override_no_updates: false
+    # list of forward DNS servers
+    # selects based on FQDN longest match
+    forward: 
+       - name: "example.com."
+         key: ""
```
Comment (astro) [inline example.yaml:196]: Empty key name looks weird. If intended, elaborate in a comment?

File: libs/ddns/src/update.rs [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+            let mut s = ip
+                .octets()
+                .iter()
+                .rev()
+                .map(|o| {
+                    // convert u8 into reverse nibbles
+                    // 1 byte fa -> "a.f"
+                    let a = char::from_digit(((*o >> 4) & 0xF) as u32, 16).unwrap();
+                    let b = char::from_digit((*o & 0xF) as u32, 16).unwrap();
+                    format!("{b}.{a}")
```
Comment (astro) [inline libs/ddns/src/update.rs:256]:
```suggestion
                    format!("{o:02x}")
```

**Review by astro — APPROVED** [review]
With ISC DHCPd at its end of life, [I went for kea in the meanwhile](https://gitea.c3d2.de/zentralwerk/network/src/branch/master/nix/nixos-module/container/dhcp-server.nix). I really like that the config formats look quite similar so we can switch easily in the future.

File: libs/ddns/src/update.rs (hunk 2 above)
Comment (leshow) [reply libs/ddns/src/update.rs:256]: this suggestion doesn't reverse the nibbles, unless I am missing something? for ex.
`188 -> c.b`
with format!("{o:02x}")
`188 -> bc`

File: example.yaml (hunk 1 above)
Comment (leshow) [reply example.yaml:196]: Yes, I can remove that from the example.yaml. The `key` field is in fact optional so it doesn't need to be present

**Comment by leshow** [comment]
@astro would you be open to help test if I clean up this branch and merge it?

**Comment by stappersg** [comment]
On Wed, Nov 15, 2023 at 12:26:02PM -0800, Evan Cameron wrote:
> +                    // convert u8 into reverse nibbles
> +                    // 1 byte fa -> "a.f"
> 
> this suggestion doesn't reverse the nibbles, unless I am missing something? for ex.
> `188 -> c.b`
> with format!("{o:02x}")
> `188 -> bc`
> 

I assume that reversing the nibbles
is for IPv6  "reverse address" a.k.a.  PTR  record

**Comment by leshow** [comment]
I've rebased off master and squashed all the commits into 1, so if anybody wants to take this for a spin, please let me know.

**Comment by leshow** [comment]
I'm going to merge this and if there are issues, feel free to open a ticket.

File: libs/ddns/src/update.rs (hunk 2 above)
Comment (astro) [reply libs/ddns/src/update.rs:256]: Oops, you are right

**Pull request merged**